File: AgentLens/Services/Search/VectorSearch/VectorCandidateBackend.swift
import Foundation
import OpenBurnBarCore

// MARK: - Vector Candidate Backend Protocol

/// Protocol for vector search backends that can find similar vectors.
protocol VectorCandidateBackend: AnyObject {
    /// Unique identifier for this backend.
    var id: String { get }

    /// Rebuilds the index with new entries.
    func rebuild(entries: [VectorIndexEntry], distanceMetric: EmbeddingDistanceMetric) throws

    /// Finds the top-k candidates most similar to the query vector.
    func candidates(for queryVector: [Float], limit: Int) throws -> [VectorIndexCandidate]
}

// MARK: - Exact Vector Backend

/// Exact k-NN search backend that scans all vectors.
/// Provides perfect accuracy but has O(n) time complexity.
final class ExactVectorCandidateBackend: VectorCandidateBackend {
    let id = "exact_scan_v1"
    private var entries: [VectorIndexEntry] = []
    private var distanceMetric: EmbeddingDistanceMetric = .cosine
    private var dimensions = 0

    func rebuild(entries: [VectorIndexEntry], distanceMetric: EmbeddingDistanceMetric) throws {
        self.entries = entries
        self.distanceMetric = distanceMetric
        self.dimensions = entries.first?.vector.count ?? 0
    }

    func candidates(for queryVector: [Float], limit: Int) throws -> [VectorIndexCandidate] {
        guard limit > 0, entries.isEmpty == false else { return [] }
        guard dimensions == 0 || queryVector.count == dimensions else {
            throw VectorIndexBackendError.dimensionMismatch(expected: dimensions, actual: queryVector.count)
        }

        var scored: [VectorIndexCandidate] = []
        scored.reserveCapacity(entries.count)
        for entry in entries {
            let score = VectorMath.similarity(lhs: queryVector, rhs: entry.vector, metric: distanceMetric)
            guard score.isFinite else { continue }
            scored.append(VectorIndexCandidate(chunkID: entry.chunkID, score: score))
        }

        scored.sort {
            if $0.score == $1.score {
                return $0.chunkID < $1.chunkID
            }
            return $0.score > $1.score
        }

        if scored.count > limit {
            return Array(scored.prefix(limit))
        }
        return scored
    }
}

// MARK: - Signpost ANN Backend

/// Approximate Nearest Neighbor backend using signpost (binary hashing) indexing.
/// Provides fast O(1) lookup with good accuracy for L2-normalized vectors.
/// Delegates to the shared `BurnBarSignpostVectorIndex` in OpenBurnBarCore.
final class SignpostANNVectorCandidateBackend: VectorCandidateBackend {
    let id = "ann_signpost_v1"

    private let index: BurnBarSignpostVectorIndex

    init(
        bucketBits: Int = 12,
        candidateMultiplier: Int = 6,
        maxHammingDistance: Int = 1
    ) {
        self.index = BurnBarSignpostVectorIndex(
            bucketBits: bucketBits,
            candidateMultiplier: candidateMultiplier,
            maxHammingDistance: maxHammingDistance
        )
    }

    func rebuild(entries: [VectorIndexEntry], distanceMetric: EmbeddingDistanceMetric) throws {
        let coreEntries = entries.map {
            BurnBarVectorIndexEntry(chunkID: $0.chunkID, vector: $0.vector)
        }
        let coreMetric: BurnBarEmbeddingDistanceMetric
        switch distanceMetric {
        case .cosine:
            coreMetric = .cosine
        case .dotProduct:
            coreMetric = .dotProduct
        case .euclidean:
            coreMetric = .euclidean
        }
        index.rebuild(entries: coreEntries, distanceMetric: coreMetric)
    }

    func candidates(for queryVector: [Float], limit: Int) throws -> [VectorIndexCandidate] {
        let coreCandidates = index.candidates(for: queryVector, limit: limit)
        return coreCandidates.map {
            VectorIndexCandidate(chunkID: $0.chunkID, score: $0.score)
        }
    }
}

File: windows/app/OpenBurnBar.App.MemorySearch/Search/VectorCandidateBackend.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace OpenBurnBar.App.MemorySearch.Search;

// PORTED (faithful) from
//   AgentLens/Services/Search/VectorSearch/VectorIndexTypes.swift        (entry/candidate/kind/error)
//   AgentLens/Services/Search/VectorSearch/VectorCandidateBackend.swift  (protocol + ExactVectorCandidateBackend)
//
// THE canonical ranking rule appears verbatim in three Swift sites
// (VectorCandidateBackend.swift:48-53, VectorSemanticProvider.swift:368-373 and 747-752):
//   drop non-finite scores → sort by score DESCENDING, tiebreak chunkID ASCENDING → take(limit).
// It is implemented once here (VectorRanking.RankTopK) and reused by the exact backend and
// the semantic exact-rerank. chunkIDs are ASCII, so `String <` == StringComparer.Ordinal.

/// <summary>Which vector backend to use. Swift: <c>enum VectorBackendKind</c>.</summary>
public enum VectorBackendKind
{
    Ann,
    Exact,
}

/// <summary>An index entry: a chunk id and its vector. Swift: <c>struct VectorIndexEntry</c>.</summary>
public sealed record VectorIndexEntry(string ChunkId, IReadOnlyList<float> Vector);

/// <summary>A scored candidate. Swift: <c>struct VectorIndexCandidate</c>.</summary>
public sealed record VectorIndexCandidate(string ChunkId, double Score);

/// <summary>Vector index error. Swift: <c>enum VectorIndexBackendError</c>.</summary>
public sealed class VectorIndexBackendException : Exception
{
    public int Expected { get; }

    public int Actual { get; }

    public VectorIndexBackendException(int expected, int actual)
        : base($"Vector dimension mismatch. Expected {expected}, got {actual}.")
    {
        Expected = expected;
        Actual = actual;
    }
}

/// <summary>
/// The single canonical top-k ranking rule shared across the Swift search code. Implemented
/// once so every ranking site is bit-identical.
/// </summary>
public static class VectorRanking
{
    /// <summary>
    /// Drop non-finite scores, sort by score DESCENDING with chunkID ASCENDING tiebreak, take
    /// the first <paramref name="limit"/>. Swift comparator: <c>if a.score == b.score { return
    /// a.chunkID &lt; b.chunkID }; return a.score &gt; b.score</c>.
    /// </summary>
    public static List<VectorIndexCandidate> RankTopK(IEnumerable<VectorIndexCandidate> scored, int limit)
    {
        ArgumentNullException.ThrowIfNull(scored);
        var finite = new List<VectorIndexCandidate>();
        foreach (var candidate in scored)
        {
            if (double.IsFinite(candidate.Score))
            {
                finite.Add(candidate);
            }
        }

        finite.Sort(Compare);
        if (limit >= 0 && finite.Count > limit)
        {
            return finite.GetRange(0, limit);
        }

        return finite;
    }

    /// <summary>The canonical comparator: score descending, then chunkID ascending (ordinal).</summary>
    public static int Compare(VectorIndexCandidate a, VectorIndexCandidate b)
    {
        ArgumentNullException.ThrowIfNull(a);
        ArgumentNullException.ThrowIfNull(b);
        if (a.Score == b.Score)
        {
            return string.CompareOrdinal(a.ChunkId, b.ChunkId);
        }

        // Descending: larger score sorts first.
        return b.Score.CompareTo(a.Score);
    }
}

/// <summary>The candidate-backend seam. Swift: <c>protocol VectorCandidateBackend</c>.</summary>
public interface IVectorCandidateBackend
{
    string Id { get; }

    void Rebuild(IReadOnlyList<VectorIndexEntry> entries, EmbeddingDistanceMetric distanceMetric);

    /// <summary>Top-<paramref name="limit"/> candidates for the query vector.</summary>
    IReadOnlyList<VectorIndexCandidate> Candidates(IReadOnlyList<float> queryVector, int limit);
}

/// <summary>
/// Deterministic exact k-NN backend. Swift: <c>ExactVectorCandidateBackend</c>
/// (<c>id = "exact_scan_v1"</c>). Scores every entry with <see cref="VectorMath.Similarity"/>,
/// drops non-finite, ranks with the canonical comparator, truncates to <c>limit</c>.
/// </summary>
public sealed class ExactVectorCandidateBackend : IVectorCandidateBackend
{
    private IReadOnlyList<VectorIndexEntry> _entries = Array.Empty<VectorIndexEntry>();
    private EmbeddingDistanceMetric _distanceMetric = EmbeddingDistanceMetric.Cosine;
    private int _dimensions;

    public string Id => "exact_scan_v1";

    public void Rebuild(IReadOnlyList<VectorIndexEntry> entries, EmbeddingDistanceMetric distanceMetric)
    {
        ArgumentNullException.ThrowIfNull(entries);
        _entries = entries;
        _distanceMetric = distanceMetric;
        _dimensions = entries.Count > 0 ? entries[0].Vector.Count : 0;
    }

    public IReadOnlyList<VectorIndexCandidate> Candidates(IReadOnlyList<float> queryVector, int limit)
    {
        ArgumentNullException.ThrowIfNull(queryVector);
        if (limit <= 0 || _entries.Count == 0)
        {
            return Array.Empty<VectorIndexCandidate>();
        }

        // Dimension guard is only enforced once the index has a known dimension.
        if (_dimensions != 0 && queryVector.Count != _dimensions)
        {
            throw new VectorIndexBackendException(_dimensions, queryVector.Count);
        }

        var scored = _entries.Select(entry =>
            new VectorIndexCandidate(entry.ChunkId, VectorMath.Similarity(queryVector, entry.Vector, _distanceMetric)));
        return VectorRanking.RankTopK(scored, limit);
    }
}
